### PR TITLE
Add support for recursive up- and downloads

### PR DESCRIPTION
**Pull request by stefanseefeld**
## Description

This PR allows the `art.artifacts.deploy()` and `art.artifacts.download()` methods to operate on directories recursively.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
        (one could argue that the existing docs don't explicitly talk about arguments types (file or directory),
         would it should at least be in the release notes)
## How has it been tested ?

I ran all existing tests (obviously), and added a new test (`test_artifacts.py::test_download_folder_success`)
to make sure downloading a directory work.

I did not add a new test to verify directory uploads.

## Checklist:

- [x] My PR is ready for prime time! Otherwise use the ["Draft PR" feature](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests)
- [x] All commits have a correct title
- [ ] Readme has been updated
- [x] Quality tests are green (see Codacy)
- [x] Automated tests are green (see pipeline)

### CODE REVIEW

File: tests/test_artifacts.py
Comment (stefanseefeld) [inline tests/test_artifacts.py:57]: The above line actually fixes a minor bug, as previously the `path` would include the repo, which is wrong.

File: pyartifactory/objects.py
Comment (anancarv) [inline pyartifactory/objects.py:772]: Can you please replace replace the variable c by `child` for more readibility ?

File: pyartifactory/objects.py
Comment (anancarv) [inline pyartifactory/objects.py:776]: Same here

File: pyartifactory/objects.py
Comment (anancarv) [inline pyartifactory/objects.py:828]: Can you add a docstring for this method. Indeed, according to `PEP 8`:
```
Docstrings are not necessary for non-public methods, but you should have a comment that describes what the method does. This comment should appear after the "def" line.
```

**Review by anancarv — CHANGES_REQUESTED** [review]

**Comment by anancarv** [comment]
Hi @stefanseefeld 
First of all, thanks for your contribution. It's a great feature !!! 
I only requested a few changes, but everything seems good to me. Once they're addressed, your PR is good for merging. I also added @nymous for having a second review. Thanks again

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py]: Done.

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py]: Done.

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:828]: I have cloned the docstring ffrom `download` and adjusted both slightly.

**Review by anancarv — DISMISSED** [review]

**Comment by stefanseefeld** [comment]
Thanks @nymous for your thorough review and test. You are entirely right ! I had confused myself and complicated the code. As you can see from my follow-up commit, I addressed the problem by changing the `local_directory_path` default value from `None` to `"."`, then removed some obsolete lines.

File: pyartifactory/objects.py
Comment (nymous) [inline pyartifactory/objects.py:824]: I think I would rather change the variable name from `file` to something else than ignore a mypy error. Here mypy is confused because there is already a `file` variable declared above (in the `for` loop) with the type `str`, so it doesn't want you to reassign it to a `BinaryIO` type.
Something like this fixes it (but you can choose another name if you want):
```suggestion
            with open(local_file_location, "rb") as file_to_upload:
                self._put(f"{artifact_path}", data=file_to_upload)
```

File: pyartifactory/objects.py
Comment (nymous) [inline pyartifactory/objects.py:813]: Maybe update this documentation to indicate that it can also accept a folder?
Like this?
```suggestion
        Deploy an artifact or a folder to an Artifactory repository.
        If deploying a file, the artifact_path MUST end with the destination artifact name.
        :param local_file_location: Location of the file or folder to deploy
        :param artifact_path: Path to the artifact in Artifactory
```

File: pyartifactory/objects.py
Comment (nymous) [inline pyartifactory/objects.py:866]: Would it be safer to use `os.makedirs(local_path, exist_ok=True)`? That way we can keep downloading in the same folder every time.

**Review by nymous — CHANGES_REQUESTED** [review]
Awesome! Sorry for the delay ^^'
You fixed the bug I found! I still have a few remarks, but after that it's good to go!

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:758]: You are right, I cloned it from `os.walk()` even though at this time it isn't being used. I was hesitating whether or not to keep it, but opted to leave it in should we ever want to need it. (For example, it might be useful at some point to add more test cases, where some files could trigger errors during attempt to download them.)
Either way, I'm not hung up on it, so if you guys would prefer not to have the unused parameter there, I can certainly remove it.

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:776]: Sure, I can do that.

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:824]: Here I actually disagree: there are actually two distinct variables (within non-overlapping scopes), so it's actually a shortcoming of the `mypy` tool not to support this -yet. (There are discussions to improve that, see https://github.com/python/mypy/issues/6232 for example.)

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:866]: I'm not sure I understand. The walk's `topdown` flag guarantees that missing parent directories will be created first, so I don't see where `os.mkdir()` would fail where `os.makedirs()` wouldn't. Or at least, if it would fail, it would already fail  prior to my PR, so I didn't add any implicit pre-conditions. :-)

File: pyartifactory/objects.py
Comment (nymous) [reply pyartifactory/objects.py:824]: Oh, I see. Too bad about mypy ^^'
I don't know how to handle it, there is a workaround to avoid bypassing the error but it's annoying. I'm afraid we will forget it there and silence too many errors without knowing it.

File: pyartifactory/objects.py
Comment (nymous) [reply pyartifactory/objects.py:866]: The issue isn't about non-existing folders, it's about already existing ones ^^
If you download a folder again for example, the folder structure is already there and `os.mkdir` errors if the directory already exists.

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:824]: I can add a comment to explain the issue. (But I do have a rather strong aversion against changing perfectly fine code just to make some tool happy ;-) )

File: pyartifactory/objects.py
Comment (stefanseefeld) [reply pyartifactory/objects.py:866]: Oh, indeed, good point ! :-)

**Review by nymous — APPROVED** [review]
With this comment about mypy I approve!
@anancarv you can merge whenever you want 😉

**Pull request merged**
